src/assembler.rs

```diff
@@ -1,15 +1,17 @@
 // SPDX-License-Identifier: (Apache-2.0 OR MIT)
 // Copyright 2017 Rich Lane <lanerl@gmail.com>
 
-//! This module translates eBPF assembly language to binary.
+//! This module translates eBPF assembly language to binary. It depends on the
+//! `combine` crate, which depends on the standard library. Because of this
+//! it is only available when the `std` feature is enabled.
```

Comment:
Looking at the amount of tests that you had to leave out when we use `std`, it's rather annoying to leave the assembler out. How much did you look into `combine`'s support for no_std? [Looking at its changelog](https://github.com/Marwes/combine/blob/f726b8f943f44d258686911907c0324771a2d09f/CHANGELOG.md?plain=1#L446), it should be able to work without `std` since v3 (we're using v4+). Could we deselect `std` for that crate when we deselect it for rbpf, keep building the assembler (and disassembler if possible), and keep the tests?

Reply:
Good point, I will try and enable it again tonight. From what I remember, simply specifying `default-features=false` in the Cargo.toml didn't cut it but given the change log that you have shared there might still be a way to use it under no_std. It could be that the alloc feature needs to be enabled. 

I do agree that quite a few tests needed to be left out and it would be good to enable those. 
As for the disassembler, this module seems to work fine with `no_std`, it is just that some of its tests require the assembler to prepare the test case which means that they need to be disabled

Reply:
The changelog does indeed look promising, I was probably just misconfiguring something. I don't think there is something that needs to happen when rbpf calls into `combine` that would require the standard library. From what I could tell it was some parser combinators stuff

Reply:
Oh, I doublechecked and unfortunately it seems like `rbpf` depends on the `EasyParser` from combine which for some reason is available on `std` only (see [here](https://docs.rs/combine/latest/combine/trait.EasyParser.html#:~:text=Available%20on%20crate%20feature%20std%20only.)

In this case there are two ways of moving forward: 
- leave `asm_parser` and assembler disabled
- change the implementation of the `asm_parser` to not depend on the `EasyParser`

From what I can tell, there is only one place where `rbpf` depends on that trait:
```rust
/// Parse a string into a list of instructions.
///
/// The instructions are not validated and may have invalid names and operand types.
pub fn parse(input: &str) -> Result<Vec<Instruction>, String> {
    match spaces()
        .with(many(instruction()).skip(eof()))
        .easy_parse(position::Stream::new(input))
    {
        Ok((insts, _)) => Ok(insts),
        Err(err) => Err(err.to_string()),
    }
}
```
According to the docs, the easy parser is supposed to produce better error messages. However in this place the error is turned directly into a string, so the code doesn't actually depend on the structue of the returned errors. I will and see if I can replace that with the normal parser. There are two possible approaches:
- replace this with the regular parser and sacrifice the nice error messagea
- instrument the code to use `EasyParser` when we have access to `std` and `Parser` otherwise

Reply:
If we end up skipping two tests only, then yes it's absolutely fine! Thanks a lot for looking into this!

Reply:
Yes, it turned out that the only difference between `EasyParser` and the `Parser` in combine is that they produce different error messages. Only these two test cases above depend on the contents of those errors, so I added some conditional `cfg` to change the expedted error depending on the enabled features. Aside from this all other tests behave the same under both environments.